File: project RAD/BFS/problem2/sol.java
// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
    }
}

public class sol {

    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();

        if (root == null)
            return result;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size(); // number of nodes at this level
            List<Integer> currentLevel = new ArrayList<>();

            // Process one level
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                currentLevel.add(node.val);

                // Add children for next level
                if (node.left != null)
                    queue.offer(node.left);
                if (node.right != null)
                    queue.offer(node.right);
            }

            // Add this level to result
            result.add(currentLevel);
        }

        return result;
    }

    public static void main(String[] args) {
        // Tree 1: [3, 9, 20, null, null, 15, 7]
        TreeNode root1 = new TreeNode(3);
        root1.left = new TreeNode(9);
        root1.right = new TreeNode(20);
        root1.right.left = new TreeNode(15);
        root1.right.right = new TreeNode(7);

        // Tree 2: [1, 2, 3, 4, null, null, 5]
        TreeNode root2 = new TreeNode(1);
        root2.left = new TreeNode(2);
        root2.left.left = new TreeNode(4);
        root2.right = new TreeNode(3);
        root2.right.right = new TreeNode(5);

        List<TreeNode> testRoots = Arrays.asList(root1, root2, null, new TreeNode(1));

        for (TreeNode root : testRoots) {
            List<List<Integer>> result = levelOrder(root);
            System.out.println("Traversal: " + result);
        }
    }
}

File: project RAD/Symmetric tree problem/Sol.py
from typing import Optional

# Conceptual Node Definition
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isSymmetric(root: Optional[TreeNode]) -> bool:
    """
    Checks if a binary tree is symmetric around its center.
    """
    if not root:
        return True
    
    # Start the mirror check on the left and right children
    return isMirror(root.left, root.right)

def isMirror(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    """
    Helper function to check if two subtrees (p and q) are mirrors of each other.
    """
    
    # Base Case 1: Both nodes are null (end of a branch), they are symmetric
    if p is None and q is None:
        return True
        
    # Base Case 2: One node is null, or their values differ. Not symmetric.
    if p is None or q is None or p.val != q.val:
        return False
        
    # Recursive Step: Check the mirror relationship:
    # 1. p's LEFT must mirror q's RIGHT.
    # 2. p's RIGHT must mirror q's LEFT.
    return isMirror(p.left, q.right) and isMirror(p.right, q.left)

# --- Test Cases ---

if __name__ == "__main__":
    
    # Example 1: Symmetric Tree (Expected: True)
    #      1
    #     / \
    #    2   2
    #   / \ / \
    #  3  4 4  3
    root1 = TreeNode(1, 
                     TreeNode(2, TreeNode(3), TreeNode(4)),
                     TreeNode(2, TreeNode(4), TreeNode(3)))
    
    # Example 2: Asymmetric Tree (Expected: False)
    #      1
    #     / \
    #    2   2
    #     \   \
    #      3   3
    root2 = TreeNode(1, 
                     TreeNode(2, None, TreeNode(3)),
                     TreeNode(2, None, TreeNode(3)))

    print("--- Symmetric Tree (Recursive DFS) Test Results ---")
    
    tests = [(root1, True), (root2, False), (None, True)]
    
    for root, expected in tests:
        result = isSymmetric(root)
        status = "PASSED" if result == expected else f"FAILED (Expected: {expected}, Got: {result})"
        
        print(f"Test -> Result: {result}, Status: {status}")
